### PR TITLE
sys/net/if_bridge: support non-INET kernels

### PR DESCRIPTION
**Pull request by llfw**

### CODE REVIEW

File: sys/net/if_bridge.c
Comment (bsdimp) [inline sys/net/if_bridge.c:3562]: These should be at the start of the block. Or at least declared there (make the case ETHERTYPE_IP: start a new block).

**Comment by llfw** [comment]
updated for C89 style.

> I was concerned about #define abuse initially

i did spend a while thinking about the best way to do this and i couldn't come up with a better method.  there is a lot of code in src (not just if_bridge) which does things like:

```
if (AF_INET code...
#ifdef INET6
    || AF_INET6 code...
#endif
    ) {
  ...
```

and it's really awkward to convert this to support INET6-only kernels.  in this case i thought the additional macros were the least awful solution, but i'm open to other suggestions.

File: sys/net/if_bridge.c
Comment (markjdb) [inline sys/net/if_bridge.c:3473]: Why is link-layer filtering conditional on INET? I'm not too familiar with this bit of the code, but it looks wrong.

File: sys/net/if_bridge.c
Comment (markjdb) [inline sys/net/if_bridge.c:3653]: IMO the ifdef should come before the comment which applies to the conditionally compiled code.

**Review by markjdb — COMMENTED** [review]
Broadly seems ok to me. I had a couple of inline comments.

I'm all for not having preprocessor conditionals in the middle of expressions. :)

File: sys/net/if_bridge.c
Comment (llfw) [reply sys/net/if_bridge.c:3473]: if we're an INET6-only kernel and the packet is an IPv4 packet, or vice versa, it will already be handled by the previous check on line 3448, which either drops the packet or unconditionally accepts it based on the value of `net.link.bridge.pfil_onlyip`, and we should never get here.

however, it does seem like there might be a pre-existing bug here when an INET6 kernel won't pass IPv6 packets to the filter.

i'm not sure under what circumstances `PFIL_HOOKED_OUT(V_link_pfil_head)` can be true while `PFIL_HOOKED_OUT(V_inet6_pfil_head)` is false -- perhaps this should be replaced by a check for `PFIL_HOOKED_OUT_ANY`?

File: sys/net/if_bridge.c
Comment (llfw) [reply sys/net/if_bridge.c]: fixed.

File: sys/net/if_bridge.c
Comment (llfw) [reply sys/net/if_bridge.c:3473]: i tested this code:

```
        if (PFIL_HOOKED_OUT_ANY && V_pfil_ipfw != 0 && dir == PFIL_OUT && ifp != NULL) {
                switch (pfil_mbuf_out(V_link_pfil_head, mp, ifp, NULL)) {
                case PFIL_DROPPED:
                        return (EACCES);
                case PFIL_CONSUMED:
                        return (0);
                }
        }
```

and it does compile, but since i don't have any systems using bridge filtering, i can't really test it.

**Comment by glebius** [comment]
I like the proposed style more than what we got, like the example given:

> ```
> if (AF_INET code...
> #ifdef INET6
>     || AF_INET6 code...
> #endif
>     ) {
>   ...
> ```

I didn't do a thorough review, but change looks good to me. I'd be glad to see it pushed. Thanks!

File: sys/net/if_bridge.c
Comment (gmshake) [inline sys/net/if_bridge.c:153]: Actually PFIL can operate on link-layer hooks, so I'd prefer `PFIL_HOOKED_IN_46` rather than `PFIL_HOOKED_IN_ANY`.

Comment (llfw) [reply sys/net/if_bridge.c]: fixed (i hope -- had some git issues locally...)

**Comment by bsdimp** [comment]
Need to verify the latest patches to see if we need it.

File: sys/net/if_bridge.c
Comment (markjdb) [reply sys/net/if_bridge.c:3473]: I have no good way to test this at the moment either. I'd prefer to see this patch split into two: one which adds preprocessor symbols that let us avoid the ugly inline ifdefs, and one for the rest of the change. The former seems uncontroversial and a positive change, but I am not sure about the latter.

**Comment by bsdimp** [comment]
I like @markjdb 's idea of splitting this. That's the change I'm flagging.

**Comment by llfw** [comment]
the uncontroversial changes have been moved #1191.  i'll update/rebase this PR once that's merged.

**Comment by llfw** [comment]
#1191 was merged (thanks Mark) so i've rebased this on current main.

based on previous discussion i believe this is the correct logic for `bridge_pfil()` (and my previous patch was wrong) but i've only determined this by code inspection, not testing.  i'll bring up a local VM here with a couple of vnet jails to test this properly.

**Comment by bsdimp** [comment]
This looks good to me... @glebius @gmshake What do you think? I'll land this if there's no objection soon.

**Comment by glebius** [comment]
LGTM

File: sys/net/if_bridge.c
Comment (gmshake) [inline sys/net/if_bridge.c:418]: No need for extra blank.

File: sys/net/if_bridge.c
Comment (gmshake) [inline sys/net/if_bridge.c:407]: These two prototype can be untouched, so the diff is smaller.

File: sys/net/if_bridge.c
Comment (gmshake) [inline sys/net/if_bridge.c:414]: No need for extra blank.

File: sys/net/if_bridge.c
Comment (gmshake) [inline sys/net/if_bridge.c:3750]: The `#ifdef #endif` block is long, and I think comment `/* INET */` after `#endif` will improve readability.

File: sys/net/if_bridge.c
Comment (gmshake) [inline sys/net/if_bridge.c:3883]: Ditto.

**Review by gmshake — APPROVED** [review]
Looks good to me ( but for some nits ).

**Comment by gmshake** [comment]
Oops, it has been merged before my review. Anyway still good :)

**Comment by emaste** [comment]
> Oops, it has been merged before my review. Anyway still good :)

Post-commit review is fine too :)

I agree with your notes, do you want to just make those changes in the tree?

**Comment by gmshake** [comment]
> > Oops, it has been merged before my review. Anyway still good :)
> 
> Post-commit review is fine too :)
> 
> I agree with your notes, do you want to just make those changes in the tree?

Committed in 73585176ffd84c13d68cad67c2ca81643f09075c .

**Comment by bsdimp** [comment]
This can be closed per subitter

**Comment by gmshake** [comment]
Hi @bsdimp , I'd like to MFC this to stable branches. Do you have plan to do that ?

**Comment by bsdimp** [comment]
Just this one orvthe series? I have no plans at tge moment... but i could make plans with lists of hashes.. or if you have a commit bit, you can.

**Comment by gmshake** [comment]
> Just this one orvthe series? 

The series indeed.

> I have no plans at tge moment... but i could make plans with lists of hashes.. or if you have a commit bit, you can.

OK, I'll do that.

**Pull request closed without merging**
